documentation/writing_tests.md

```diff
@@ -0,0 +1,1473 @@
+
+```ts
+describe("Tests for the internal classes", function() {
+
+  describe("Tests for the InternalClass", function() {
+    it("The InternalClass should be able to be initialized", function() {
+      const Internal = new InternalClass();
+      assert.exists(Internal);
+    });
+
+    it("The InternalClass's returnsTrue should return true", function() {
+      const Internal = new InternalClass();
+      const result: boolean = Internal.returnsTrue();
+      assert.isTrue(result);
+    });
+  });
+
+  describe("Tests for the InternalInheritedClass", function() {
+    it("The InternalInheritedClass should be able to be initialized", function() {
+      const Internal = new InternalInheritedClass();
+      assert.exists(Internal);
+    });
+
+    // We're not testing returnsTrue because we've already tested the parent class's returnsTrue.
+
+    it("The InternalInheritedClass's returnsFalse should return false", function() {
+      const Internal = new InternalInheritedClass();
+      const result: boolean = Internal.returnsFalse();
+      assert.isFalse(result);
+    });
+  });
+
+});
+```
+
+Keep in mind that **We do not test inherited members unless we do something that uniquely changes the way they act**.
+
+In addition to being expressive, test titles might have specific hashtags (as in, words preceded by the pound sign), to facilitate pattern matching.
+
+You can see a working example of tests for internal code in our [test/internal/ folder in our template project](../sdk/template/template/test/internal/).
+
+#### Pattern matching test titles
+
+Mocha allows us to only run certain tests as long as they match a given query, with the following arguments:
+
+```
+Test Filters
+  --fgrep, -f   Only run tests containing this string                   [string]
+  --grep, -g    Only run tests matching this string or regexp           [string]
+  --invert, -i  Inverts --grep and --fgrep matches                     [boolean]
+```
+
+To be able to filter tests through pattern matching with any of these filters, the `describe` and `it` titles can have any hashtag, but preferably one or more of the following:
+
+- `#browser` for tests that are only expected to run in the browser.
+- `#node` for tests that are only expected to run on NodeJS.
+- `#live` for tests that should only run live and not be recorded, nor played back.
+
+These and any other hashtags can be used to provide custom `package.json` scripts to run specific sets of tests. We still recommend [using conditionals](#using-conditionals), to ensure that someone reading any test case clearly understands under what conditions this test will be executed or skipped.
```

Comment:
Other than just about the mocha flags, you might want to add about adding mocha config in karma.conf too to get the same behavior for browser tests
[snippet from karma.conf](https://github.com/HarshaNalluru/azure-sdk-for-js/blob/85a55ab75734e01b1b12f9112eb7611b03a794c1/sdk/eventhub/event-hubs/karma.conf.js#L117-L123)

Reply:
Thank you!

Reply:
It should be fine now!